src/encoder.rs

```diff
@@ -1137,187 +1137,201 @@ pub fn save_block_motion<T: Pixel>(
   }
 }
 
-pub fn encode_block_a<T: Pixel>(
-  seq: &Sequence, fs: &FrameState<T>,
-  cw: &mut ContextWriter, w: &mut dyn Writer,
-  bsize: BlockSize, bo: BlockOffset, skip: bool
-) -> bool {
-  cw.bc.blocks.set_skip(bo, bsize, skip);
-  if fs.segmentation.enabled && fs.segmentation.update_map && fs.segmentation.preskip {
-    cw.write_segmentation(w, bo, bsize, false, fs.segmentation.last_active_segid);
-  }
-  cw.write_skip(w, bo, skip);
-  if fs.segmentation.enabled && fs.segmentation.update_map && !fs.segmentation.preskip {
-    cw.write_segmentation(w, bo, bsize, skip, fs.segmentation.last_active_segid);
-  }
-  if !skip && seq.enable_cdef {
-    cw.bc.cdef_coded = true;
-  }
-  cw.bc.cdef_coded
+pub struct BlockEncoder<'a, T: Pixel> {
```

Comment:
I attempted to rebase my local branch about tiling on this PR, but this structure adds a lot of constraints on lifetimes.

In particular, I use structs with lifetimes, so I replace:
 - `&'a mut ContextWriter` by `&'a mut ContextWriter<'a>`
 - `&'a mut FrameState<T>` by `&'a mut TileStateMut<'a, T>`

The lifetimes constraints cannot be expressed on closure parameters:

```rust
struct X<'a>(&'a mut i32, &'a mut i32);

fn main() {
    fn function<'a>(a: &'a mut i32, b: &'a mut i32) -> X<'a> { X(a, b) }
    let lambda = |a: &mut i32, b: &mut i32| { X(a, b) };
    let mut a = 0;
    let mut b = 1;
    let x = function(&mut a, &mut b);
    // let x = lambda(&mut a, &mut b);
}

It happens to work on this example if we remove the explicit labels:
```
let lambda = |a, b| { X(a,b) }
```
but this works only if the compiler is able to infer them properly.

Therefore, with this change, I need to convert this [closure](https://github.com/xiph/rav1e/blob/master/src/rdo.rs#L582) (which itself contains another closure which should probably be converted too) to a function (which, in itself, is not a bad idea). But then, it does not capture the variables from the environment anymore, so I also need to add `&'a FrameInvariants<T>` and other fields.

But then, all the lifetimes are too constrained.

```rust
error[E0495]: cannot infer an appropriate lifetime for lifetime parameter 'a in function call due to conflicting requirements
   --> src/rdo.rs:716:5
    |
716 | /     luma_chroma_mode_rdo(fi, luma_mode, ts, cw, bsize, bo, &mut best, mvs, ref_frames_set[i], rdo_type,
717 | |              &mode_set_chroma, false, mode_contexts[i], &mv_stacks[i]);
    | |______________________________________________________________________^
    |
note: first, the lifetime cannot outlive the lifetime  as defined on the body at 694:28...
   --> src/rdo.rs:694:28
    |
694 |   mode_set.iter().for_each(|&(luma_mode, i)| {
    |                            ^^^^^^^^^^^^^^^^^
note: ...so that reference does not outlive borrowed content
   --> src/rdo.rs:716:41
    |
716 |     luma_chroma_mode_rdo(fi, luma_mode, ts, cw, bsize, bo, &mut best, mvs, ref_frames_set[i], rdo_type,
    |                                         ^^
note: but, the lifetime must be valid for the lifetime 'a as defined on the function body at 448:26...
   --> src/rdo.rs:448:26
    |
448 | pub fn rdo_mode_decision<'a, T: Pixel>(
    |                          ^^
    = note: ...so that the expression is assignable:
            expected &mut tiling::tile_state::TileStateMut<'_, T>
               found &mut tiling::tile_state::TileStateMut<'a, T>
```
(and many more)

I cannot do what the compiler suggests, since the structure requires all the lifetimes to be the same (`'a`).

One solution to relax these constraints would be to use different lifetimes for all fields, but then that's very verbose:
```rust
pub struct BlockEncoder<'a, 'b, 'c, 'd', 'e, 'f, … T: Pixel> { … }
```

So in practice, it adds a lot of complexity.

Note that I also choose to use the same lifetime on my PR #1162 (https://github.com/xiph/rav1e/pull/1162/files#diff-e0934bb6f70a4f8c3baf22a773c14ec7R1908), but it does not cause much problems in practice (just to constraint the lifetimes of one fonction).

Reply:
Could this complexity be avoided by removing everything that's framewide from the struct? (e.g. everything with 'a)

Reply:
So all references? It would defeat the purpose of this structure I think.